File: src/bot/security/antiRaid.js
import { PermissionFlagsBits, ChannelType } from "discord.js";
import { recordStat, getConfig, getTrustLevels, pushEvent } from "../../db/database.js";
import { isTrusted } from "./trustLevels.js";

export class AntiRaid {
  constructor(client, logger) {
    this.client = client;
    this.logger = logger;
    this.joinBuckets = new Map();
    this.lockdownState = new Map();
  }

  handleMemberJoin(member) {
    const config = getConfig();
    const guildId = member.guild.id;
    const now = Date.now();
    const bucket = this.joinBuckets.get(guildId) || [];
    const fresh = bucket.filter((ts) => now - ts < 60 * 1000);
    fresh.push(now);
    this.joinBuckets.set(guildId, fresh);

    const accountAgeDays = (now - member.user.createdTimestamp) / (1000 * 60 * 60 * 24);
    const isYoung = accountAgeDays < config.raid.accountAgeDays;

    if (fresh.length >= config.raid.joinThreshold) {
      recordStat("raidAlerts");
      pushEvent({
        type: "raid",
        guildId,
        message: `Raid suspicion: ${fresh.length} joins in 60s. Account age < ${config.raid.accountAgeDays} days? ${isYoung}`
      });
      this.logger(guildId, `⚠️ Raid detected: ${fresh.length} joins in 60s. Young account: ${isYoung ? "yes" : "no"}.`);
      if (config.raid.lockdownOnRaid) {
        this.enableLockdown(member.guild, "Automatic raid lockdown");
      }
    }

    if (isYoung) {
      this.logger(guildId, `👤 New account ${member.user.tag} is ${accountAgeDays.toFixed(1)} days old.`);
      if (config.raid.kickYoungAccounts) {
        member.kick("Account too new during raid window").catch(() => {});
      } else if (config.raid.quarantineRoleId) {
        member.roles.add(config.raid.quarantineRoleId).catch(() => {});
      }
    }
  }

  async enableLockdown(guild, reason = "Lockdown enabled") {
    if (this.lockdownState.has(guild.id)) return;
    const previous = [];
    for (const channel of guild.channels.cache.values()) {
      if (channel.type !== ChannelType.GuildText) continue;
      const perm = channel.permissionOverwrites.cache.get(guild.roles.everyone.id);
      previous.push({ channelId: channel.id, allow: perm?.allow?.bitfield ?? 0n, deny: perm?.deny?.bitfield ?? 0n });
      await channel.permissionOverwrites.edit(guild.roles.everyone, {
        SendMessages: false,
        AddReactions: false
      }).catch(() => {});
    }
    this.lockdownState.set(guild.id, previous);
    this.logger(guild.id, `🚨 Lockdown enabled. Reason: ${reason}`);
  }

  async disableLockdown(guild, reason = "Lockdown disabled") {
    const previous = this.lockdownState.get(guild.id);
    if (!previous) return;
    for (const entry of previous) {
      const channel = guild.channels.cache.get(entry.channelId);
      if (!channel) continue;
      await channel.permissionOverwrites.edit(guild.roles.everyone, {
        allow: entry.allow,
        deny: entry.deny
      }).catch(() => {});
    }
    this.lockdownState.delete(guild.id);
    this.logger(guild.id, `✅ Lockdown disabled. Reason: ${reason}`);
  }

  async handleLockdownCommand(interaction, enable) {
    if (!interaction.isChatInputCommand()) return;
    const trust = getTrustLevels();
    const executorId = interaction.user.id;
    if (!isTrusted(executorId, trust)) {
      await interaction.reply({ content: "You are not trusted to run this command.", ephemeral: true });
      return;
    }

    if (enable) {
      await this.enableLockdown(interaction.guild, `Manual toggle by ${interaction.user.tag}`);
    } else {
      await this.disableLockdown(interaction.guild, `Manual toggle by ${interaction.user.tag}`);
    }
    await interaction.reply({ content: `Lockdown ${enable ? "enabled" : "disabled"}.`, ephemeral: true });
  }
}

File: bot/anti_raid.py
import datetime
from collections import defaultdict
from typing import Any

import discord

from bot.trust_levels import is_trusted


class AntiRaid:
    def __init__(self, bot: discord.Client, config: dict[str, Any]):
        self.bot = bot
        self.join_buckets: dict[int, list[datetime.datetime]] = defaultdict(list)
        self.lockdown_state: dict[int, list[tuple[int, dict[str, discord.PermissionOverwrite]]]] = {}
        self.update_config(config)

    def update_config(self, config: dict[str, Any]) -> None:
        self.config = (config or {}).get('raid', {})

    def handle_member_join(self, member: discord.Member):
        conf = self.config or {}
        now = datetime.datetime.utcnow()
        bucket = self.join_buckets[member.guild.id]
        bucket.append(now)
        cutoff = now - datetime.timedelta(seconds=60)
        self.join_buckets[member.guild.id] = [ts for ts in bucket if ts >= cutoff]
        if len(self.join_buckets[member.guild.id]) >= conf.get('joinThreshold', 10):
            if conf.get('lockdownOnRaid', True):
                self.bot.loop.create_task(self.enable_lockdown(member.guild, 'Raid detection'))
        account_age_days = (now - member.created_at.replace(tzinfo=None)).days
        if account_age_days < conf.get('accountAgeDays', 7):
            if conf.get('kickYoungAccounts', False):
                self.bot.loop.create_task(member.kick(reason='Anti-raid: account too new'))
            elif conf.get('quarantineRoleId'):
                role = member.guild.get_role(int(conf['quarantineRoleId']))
                if role:
                    self.bot.loop.create_task(member.add_roles(role, reason='Anti-raid quarantine'))

    async def enable_lockdown(self, guild: discord.Guild, reason: str):
        if guild.id in self.lockdown_state:
            return
        self.lockdown_state[guild.id] = []
        for channel in guild.text_channels:
            overwrites = channel.overwrites_for(guild.default_role)
            self.lockdown_state[guild.id].append((channel.id, {guild.default_role.id: overwrites}))
            try:
                await channel.set_permissions(guild.default_role, send_messages=False, reason=reason)
            except Exception:
                continue

    async def disable_lockdown(self, guild: discord.Guild, reason: str):
        state = self.lockdown_state.get(guild.id)
        if not state:
            return
        for channel_id, permissions in state:
            channel = guild.get_channel(channel_id)
            if not isinstance(channel, discord.TextChannel):
                continue
            overwrite = permissions.get(guild.default_role.id)
            if overwrite is None:
                continue
            try:
                await channel.set_permissions(guild.default_role, overwrite=overwrite, reason=reason)
            except Exception:
                continue
        self.lockdown_state.pop(guild.id, None)

    async def handle_lockdown_command(self, interaction: discord.Interaction, enable: bool):
        if not is_trusted(str(interaction.user.id), interaction.guild):
            await interaction.response.send_message('Action non autorisée.', ephemeral=True)
            return
        if enable:
            await self.enable_lockdown(interaction.guild, 'Manual lockdown command')
            await interaction.response.send_message('Lockdown activé.', ephemeral=True)
        else:
            await self.disable_lockdown(interaction.guild, 'Manual lockdown command')
            await interaction.response.send_message('Lockdown désactivé.', ephemeral=True)
